pkg/storegateway/series_refs.go

```diff
@@ -830,69 +832,80 @@ func metasToChunks(blockID ulid.ULID, metas []chunks.Meta) []seriesChunkRef {
-	if err := decodeSnappyGob(data, &entry); err != nil {
+	data, err := snappy.Decode(nil, data)
+	if err != nil {
+		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "can't decode series cache snappy", "err", err)
+		return seriesChunkRefsSet{}, false
+	}
+	var entry storepb.CachedSeries
+	if err = entry.Unmarshal(data); err != nil {
 		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "can't decode series cache", "err", err)
 		return seriesChunkRefsSet{}, false
 	}
-	if entry.MatchersKey != matchersKey {
+
+	if entry.MatchersKey != string(matchersKey) {
 		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "cached series entry key doesn't match, possible collision", "cached_matchers", entry.MatchersKey)
 		return seriesChunkRefsSet{}, false
 	}
-	if entry.Shard != maybeNilShard(shard) {
-		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "cached series shard doesn't match, possible collision", "cached_shard", entry.Shard)
+	if nonNilShard := maybeNilShard(shard); entry.ShardIndex != nonNilShard.ShardIndex || entry.ShardCount != nonNilShard.ShardCount {
+		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "cached series shard doesn't match, possible collision", "cached_shard_index", entry.ShardIndex, "cached_shard_count", entry.ShardCount)
 		return seriesChunkRefsSet{}, false
 	}
 
 	// This can be released by the caller because loadingSeriesChunkRefsSetIterator (where this function is called) doesn't retain it
 	// after Next() will be called again.
-	res := newSeriesChunkRefsSet(len(entry.LabelSets), true)
-	for _, lset := range entry.LabelSets {
+	res := newSeriesChunkRefsSet(len(entry.Series), true)
+	for _, lset := range entry.Series {
 		res.series = append(res.series, seriesChunkRefs{
-			lset: lset,
+			lset: mimirpb.FromLabelAdaptersToLabels(lset.Labels),
 		})
 	}
 	return res, true
 }
 
 func storeCachedSeriesForPostings(ctx context.Context, indexCache indexcache.IndexCache, userID string, blockID ulid.ULID, matchers []*labels.Matcher, shard *sharding.ShardSelector, postingsKey indexcache.PostingsKey, set seriesChunkRefsSet, logger log.Logger) {
-	entry := seriesForPostingsCacheEntry{
-		LabelSets:   make([]labels.Labels, set.len()),
-		MatchersKey: indexcache.CanonicalLabelMatchersKey(matchers),
-		Shard:       maybeNilShard(shard),
+	nonNilShard := maybeNilShard(shard)
+	matchersKey := indexcache.CanonicalLabelMatchersKey(matchers)
+	data, err := encodeCachedSeriesForPostings(set, matchersKey, nonNilShard)
+	if err != nil {
+		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "can't encode series for caching", "err", err)
+		return
+	}
+	indexCache.StoreSeriesForPostings(ctx, userID, blockID, matchersKey, shard, postingsKey, data)
+}
```

Comment:
Not directly related to this PR, but I guess this method was created in the streaming store-gateway implementation PRs?

Why does it need all three `matchersKey`, `shard` and `postingsKey`? The first two reference the same data set as the third one, aren't they?

Reply:
I'm going to move this comment to the original PR: https://github.com/grafana/mimir/pull/3687

Reply:
I think this is outside of scope of this PR, but it's in scope for what i was going to do next.

Your comment made me realize that caching matchers may even introduce false-negative cache misses. You can reword a matcher to select the same series. 
```
{a="1"}
and
{a!="2", a!="3"}
```
_may_ still select the same postings, but we'll cache them separately. 

I think the case is different for the shard because we include it in the cache key verbatim, not hash it - that reduces the collisions to only sets that have the same shard. If we also hash the shard key, then cache keys for all shards can collide.


What I'm unsure about is the strength of the hash if we don't hash the matchers - if we hash a matcher and a set of postings, is it the same strength as hashing a set of postings? Given that the matchers + shard have a 1:1 relationship with a set of postings, my answer would be no.

We'll be hashing a smaller thing - only set of postings, so the set of inputs that map onto 52 bits of hash will be smaller.. should give each input value a slightly smaller chance of collisions. So removing the matchers should even make the hash stronger?

Reply:
The method name says "StoreSeriesForPostings" so you should just store series for `postingsKey`. You don't care about matchers, shards or whatever else led you to look up those postings. Since the cache key will be the a hash of the postings, you should store the postings list itself in the cached value too, to verify that you brought the expected cached item when retrieving.

Reply:
> you should store the postings list itself in the cached value too, to verify that you brought the expected cached item when retrieving

in this case we'll store postings twice - once for expanded postings and once for the series. I wanted to avoid that. WDYT?

Reply:
an alternative is to store the number of postings

Reply:
Okay, I see your concern now. However, when encoded with delta encoding, most of the postings are less than couple of bytes, so storing them alongside the series shouldn't be a concern (because series are potentially hundreds of bytes).

(OTOH, _number of postings_ is just another hashing function, one of the bad ones :smile: )

Reply:
i know why we need the shard - the postings we have are not necessarily postings that belong only to this shard - see this comment 

https://github.com/grafana/mimir/blob/73e834e0c6a99dff1dee15e0525036bd9ec57f20/pkg/storegateway/series_refs.go#L725-L726

this means that the postings for different shards are actually the same. But then the series that we cache are filtered by the shard. So we definitely need the shard in the cache key.

But I agree we can remove the matchers.

---

on verifying the postings for collisions.. Are you aware that we currently store two checksums of the postings - blake32 and sha1?